bootcompiler/src/main/scala/org/mozartoz/bootcompiler/transform/OnStackMarking.scala

```diff
@@ -0,0 +1,221 @@
+ * by precaution, all captured variable that were not bound should be instantiated
+ * 	This means that any A = {IntToFloat 3} will instantiate because it results in {IntToFloat 3 A}
+ *  that could potentially call end up calling A
+ *  
+ * When having two branches, if a variable does not end in the same state, it it better to instantiate it
+ * 
+ * If a variable is encountered, it is always put into the modified set (if any) except if it was already
+ * in a BoundFirst or a ToInstantiate state.
+ * 
+ * In try-catch or andthen/orelse nodes, variables that have not been bound yet end up in ToInstantiate state
+ *  
+ * The basic idea is that if no usage appears before binding, no need to create a variable.
+ */
+class OnStackMarking(val declVars: mutable.HashMap[Symbol, VarAndState],
+    val modified: Option[mutable.Set[Variable]]) extends Transformer with Walker {
+  def this(declVars: mutable.HashMap[Symbol, VarAndState]) = this(declVars, None)
+  def this() = this(mutable.HashMap())
+  
+  def tagVariable(v: Variable, state: VarState) = {
+    declVars.put(v.symbol, VarAndState(v, state))
+    modified.foreach { set => set.add(v) }
+  }
+  
+  /**
+   * Looks for affected variables using a sandboxed marker
+   */
+  def affectedVariables(affectedSet: mutable.Set[Variable], marking: (OnStackMarking)=>Unit): mutable.Set[Variable] = {
+    val affected = if (affectedSet == null) mutable.Set[Variable]() else affectedSet
+    val marker = new OnStackMarking(declVars.clone(), Some(affected)) {
+      // do nothing when encountering a "bond on stack" node. Bound and Used variables will get the same treatment
+      override def onStackBind(bind: BindCommon, v: Variable) = {}
+    }
+    marking(marker)
+    affected
+  }
+  
+  /**
+   * Runs the algorithm in sandboxed branches and merges the result in current marker
+   */
+  def conjugate[A](collector: mutable.HashMap[Symbol, VarAndState], branches: ((OnStackMarking)=>Unit)*): Unit = {
+    val collectorClone = collector.clone()
+    val modified = mutable.Set[Variable]()
+    val bindNodes = mutable.HashMap[Symbol, Seq[BindCommon]]()
+    for ((branch, i) <- branches.zipWithIndex) {
+      val current = if (i == 0) collector else collectorClone.clone() 
+    	val marker = new OnStackMarking(current, Some(modified)) {
+        // In the sandboxed environment, we collect "bind on stack" nodes to defer decision
+    	  override def onStackBind(bind: BindCommon, v: Variable) =
+    	    bindNodes.put(v.symbol, bindNodes.getOrElse(v.symbol, Seq()) :+ bind)
+    	}
+      branch(marker)
+      if (i != 0) {
+        for (v <- modified) {
+          val collectorVS = collector.get(v.symbol).get
+          val currentVS = current.get(v.symbol).get
+          if (collectorVS.state != currentVS.state && collectorVS.state != ToInstantiate) {
+            collector.put(v.symbol, VarAndState(collectorVS.v, ToInstantiate))
```

Comment:
Declared and BoundFirst mean at the end of the block, we do not know whether the variable is bound or not. However, Let us take the following example:
```oz
local A in
if X then
  A = true
else
  skip
end
A = false
end
```
We absolutely have to make the second unification a true one, or more generally, any further use must have a variable to work with.
The idea is that basically if two states are different (except maybe declared and captured that could be merged into captured), it will be necessary to merge them into ToInstantiate.
BoundFirst - (Declared/Captured/ToInstantiate) => Cannot make further binding on stack! (=> ToInstantiate)
Declared - ToInstantiate => idem
Captured - ToInstantiate => idem
Declared - Captured => Here, we could keep it as Captured.